### PR TITLE
Fix Stream.junk on channel streams

### PR DESCRIPTION
**Pull request by nojb**
See [MPR#7769](https://caml.inria.fr/mantis/view.php?id=7769).

A bug happens when doing `Stream.junk` on a channel stream when the buffer is empty: `Stream.junk` never refills the buffer and simply increases the (empty) buffer index `b.ind`. The next time `Stream.next` is called, the buffer *is* refilled and the index is reset to `0`, forgetting that `Stream.junk` was called.

Rather, we should refill the buffer if needed in `Stream.junk`.

### CODE REVIEW

File: stdlib/stream.ml [hunk 1]
```diff
@@ -99,7 +99,10 @@ let rec junk_data : type v. v cell -> unit = fun s ->
   match s.data with
     Scons (_, d) -> s.count <- (succ s.count); s.data <- d
   | Sgen ({curr = Some _} as g) -> s.count <- (succ s.count); g.curr <- None
-  | Sbuffio b -> s.count <- (succ s.count); b.ind <- succ b.ind
+  | Sbuffio b ->
+      if b.ind >= b.len then fill_buff b;
+      if b.len == 0 then s.data <- Sempty
```
Comment (gasche) [inline stdlib/stream.ml:104]: I suppose this is a typo for `b.len = 0`.

**Review by gasche — APPROVED** [review]
I believe that the fix is correct, modulo a minor comment.

(I'm positively surprised by how much cleaner the Stream codebase is than the last time I looked at it. Thanks @chambart!)

(We could now use inline records to optimize the `'a t` representation, but I don't know if we still have performance-needy Stream users.)

File: stdlib/stream.ml (hunk 1 above)
Comment (nojb) [reply stdlib/stream.ml:104]: Not really -- I used the style present in the rest of the file.

Comment (gasche) [reply stdlib/stream.ml:104]: Oh, sorry.

**Comment by gasche** [comment]
@nojb feel free to cherry-pick in 4.07.

(Cherry-picking a multi-commit change is kind of tricky, good luck with git. I can do it myself if you wish. My approach is to create a copy of the branch with `-4.07` in the name, then do a `git rebase --onto 4.07 trunk <branchname>`, then merge that into 4.07.)

**Comment by nojb** [comment]
I will give it a try, thanks!

**Comment by nojb** [comment]
This is 20a87a876a32...6245f19454b2 on 4.07.

**Comment by nojb** [comment]
By the way, if your version of `git` is not too old, it is not hard to cherry-pick multiple commits. This is what I did to cherry pick this PR into 4.07:
```bash
$ git checkout 4.07
$ git cherry-pick trunk^1..fix_stream_bug
```
where `trunk^1` is the commit on `trunk` just "before" the merge of the branch `fix_stream_bug` (if you do not put in the `^1`, the commit range is empty due to `fix_stream_bug` having already been merged into `trunk`).

**Pull request merged**